packages/supervisor/README.md

```diff
@@ -0,0 +1,79 @@
+# @effection/supervisor
+
+A task supervisor similar to Erlang's OTP supervisor module. Supervisors can
+monitor other tasks and restart them as needed.
+
+The functionality and options of the Effection supervisor are broadly similar
+to Erlang's.
+
+## Known limitations:
+
+Some functionality from Erlang's supervisors is as of yet still missing:
+
+- `simple_one_for_one` supervision strategy
+- the `intensity` and `period` options, which prevent restart loops
+
+Contributions for these are very welcome!
+
+## Usage
+
+This packages exposes two methods `runSupervisor` and `createSupervisor`.
+
+With `runSupervisor`, the supervisor runs in the foreground as a task, which
+makes it easy to start and keep it running, for example with `main`:
+
+```typescript
+import { main } from 'effection';
+import { runSupervisor } from '@effection/supervisor';
+import { fobrulator, phaseInverter } from './operations';
+
+main(runSupervisor([
+  { name: 'fobrulator', run: fobrulator, args: ['fob'] },
+  { name: 'phaseInverter', run: phaseInverter, args: [true] },
+]);
+```
+
+With `createSupervisor`, the supervisor runs in the background as a resource,
+which makes it possible to dynamically control the supervisor:
+
+```typescript
+import { main } from 'effection';
+import { runSupervisor } from '@effection/supervisor';
+import { fobrulator, phaseInverter } from './operations';
+
+main(function*() {
+  let supervisor = yield createSupervisor();
+
+  supervisor.addChild({ name: 'fobrulator', run: fobrulator, args: ['fob'] });
+  supervisor.addChild({ name: 'phaseInverter', run: phaseInverter, args: [true] });
```

Comment:
What is the value in passing the args in an array rather than just doing it in a callback?

```
supervisor.addChild(({ name: 'phaseInverter', run: () => phaseInverter(true) }))
```

Reply:
That's a fair point! I basically just translated this from the Erlang version, but with anonymous closures, it is far less useful, as you say. Also, this makes it much easier to get the types right.

Reply:
Okay, looking at this again, I do see one point to this, which is that a (hypothetical) simple_one_for_one supervisor, would probably want to specify the operation to run and its arguments separately, at least that is how the Erlang version works, so it could look something like this:

```typescript
let supervisor = createDynamicSupervisor({ run: someOperation });
supervisor.addChild({ args: ['foo', 'bar', 'baz'] });
```

Or soemthing. I'm not sure what the API should look like exactly. This would mean that having `args` on the regular supervisor would add more symmetry. But it's probably a moot point, especially since we haven't even added this type of supervisor yet. What do you think, should we remove `args` for now?

Reply:
Could we have a strongly typed API for this hypothetical dynamic supervisor derived from the type of `someOperation`? For example, let's say the type of the `run` property was:

```ts
interface Run<TArgs extends unknown[]> {
  (... args: TArgs): Operation<void>;
}
```

Then we could make the create function return something like:

```ts
interface SimpleOneForOne<TArgs extends unknown[]> {
  addChild(...args: TArgs): void;
}
```

e.g.
```ts
let supervisor = createDynamicSupervisor({ run: sleep });

 // @ts-expect-error
supervisor.addChild('one hundred');

// ok!
supervisor.addChild(100)
```

If that is the case, then I think we can do without `args` in all cases.